control_node/control_node/calibration.py

```diff
@@ -0,0 +1,182 @@
+
+    def calibrate(self):
+        """Use image correspondences to compute the transformation matrix from camera to laser
+
+        1. Shoot the laser at predetermined points
+        2. For each laser point, capture an image from the camera
+        3. Identify the corresponding point in the camera frame
+        4. Compute the transformation matrix from camera to laser
+        """
+
+        # TODO: Depth values are noisy when the laser is on. Figure out how to reduce noise, or
+        # use a depth frame obtained when the laser is off
+
+        # Get calibration points
+        x_bounds = [float("inf"), float("-inf")]
+        y_bounds = [float("inf"), float("-inf")]
+        laser_bounds = self.laser.get_bounds(1.0)
+        for x, y in laser_bounds:
+            if x < x_bounds[0]:
+                x_bounds[0] = x
+            if x > x_bounds[1]:
+                x_bounds[1] = x
+            if y < y_bounds[0]:
+                y_bounds[0] = y
+            if y > y_bounds[1]:
+                y_bounds[1] = y
+
+        grid_size = (5, 5)
+        x_step = (x_bounds[1] - x_bounds[0]) / (grid_size[0] - 1)
+        y_step = (y_bounds[1] - y_bounds[0]) / (grid_size[1] - 1)
+        pending_calibration_laser_pixels = []
+        for i in range(grid_size[0]):
+            for j in range(grid_size[1]):
+                x = x_bounds[0] + i * x_step
+                y = y_bounds[0] + j * y_step
+                pending_calibration_laser_pixels.append((x, y))
+
+        # Get image correspondences
+        self.calibration_laser_pixels = []
+        self.calibration_camera_points = []
+
+        # TODO: set exposure on camera node automatically when detecting laser
+        self.camera.set_exposure(0.001)
+
+        for laser_pixel in pending_calibration_laser_pixels:
+            self.laser.stop_laser()
+            self.laser.start_laser(point=laser_pixel, color=self.laser_color)
+            # Wait for galvo to settle and for camera frame capture
+            time.sleep(0.2)
+            self.add_point_correspondence(laser_pixel)
+
+        self.laser.stop_laser()
+        self.camera.set_exposure(-1.0)
+
+        if self.logger:
+            self.logger.info(
+                f"{len(self.calibration_laser_pixels)} out of {len(pending_calibration_laser_pixels)} point correspondences found."
+            )
+
+        if len(self.calibration_laser_pixels) < 3:
+            if self.logger:
+                self.logger.warning(
+                    "Calibration failed: insufficient point correspondences found."
+                )
+            return False
+
+        # Use least squares for an initial estimate, then use bundle adjustment to refine
+        self._update_transform_least_squares()
+        self.update_transform_bundle_adjustment()
+        self.is_calibrated = True
+
+        return True
+
+    def camera_point_to_laser_pixel(self, camera_point):
+        homogeneous_camera_point = np.hstack((camera_point, 1))
```

Comment:
Honestly, this might have been causing error when I was working on the depth stuff that led to me not using the depth alignment. Looks great